servers/src/main/java/tachyon/master/file/FileSystemMaster.java

```diff
@@ -450,25 +374,47 @@ public void completeFile(long fileId)
         }
       }
 
-      completeFileInternal(fileInode.getBlockIds(), fileId, fileLength, false, opTimeMs);
+      // If the file is both persisted and stored in memory, the memory footprint and the UFS file
+      // length should match.
+      if (fileInode.isPersisted() && inMemoryLength != 0 && options.getLength() != inMemoryLength) {
+        throw new SuspectedFileSizeException("Inconsistent file length: Tachyon " + inMemoryLength
+            + " UFS " + options.getLength());
+      }
+
+      // If the file is persisted, its length is determined by UFS. Otherwise, its length is
+      // determined by its memory footprint.
+      long length = fileInode.isPersisted() ? options.getLength() : inMemoryLength;
```

Comment:
I agree. The logic will be more complicated when async persistence comes to play. Perhaps we just make under storage type explicit in options.